add json guide (#1131)
urbit/urbit.org: content/docs/hoon/guides/json-guide.md

Here's a simple `mark` file for the `$user` structure we created in the [previous section](#json-encoding-and-decoding-example). It imports the [json-test.hoon](#json-testhoon) library we created and saved in our `%home` desk's `/lib` directory.

#### `user.hoon`

```hoon
/+  *json-test
|_  usr=user
++  grab
  |%
  ++  noun  user
  ++  json  from-js
  --
++  grow
  |%
  ++  noun  usr
  ++  json  (to-js usr)
  --
++  grad  %noun
--
```

The [Marks section](/docs/arvo/clay/marks/marks) of the Clay documentation covers `mark` files comprehensively and is worth reading through if one wants to write a mark file.

In brief, a mark file contains a `door` with three arms. The door's sample type is the type of the data in question - in our case the `$user` structure. The `+grab` arm contains methods for converting _to_ our mark, and the `+grow` arm contains methods for converting _from_ our mark. The `+noun` arms are mandatory, and then we've added `+json` arms which respectively call the `+from-js` and `+to-js` functions from our `json-test.hoon` library. The final `+grad` arm defines various revision control functions, in our case we've delegated these to the `%noun` mark.

> From this mark file, Clay can build mark conversion `gate`s between the `%json` mark and our `%user` mark, allowing the conversion of `$json` data to a `$user` structure and vice versa.

Reviewer: ```suggestion
From this mark file, Clay can build mark conversion gates between the `%json` mark and our `%user` mark, allowing the conversion of `$json` data to a `$user` structure and vice versa.
```

### Try it out

First, we'll save the code above as `user.hoon` in the `/mar` directory our of `%home` desk:

```
> |commit %home
>=
+ /~zod/home/9/mar/user/hoon
```

Let's quickly create a `$json` object to work with:

```
> =jon (need (de-json:html '{"joined":1631440078,"username":"john456","name":["John","William","Smith"],"email":"john.smith@example.com"}'))
> jon
[ %o
    p
  { [p='email' q=[%s p='john.smith@example.com']]
    [p='name' q=[%a p=~[[%s p='John'] [%s p='William'] [%s p='Smith']]]]
    [p='username' q=[%s p='john456']]
    [p='joined' q=[%n p=~.1631440078]]
  }
]
```

> We'll also build our library so we can use its types from the Dojo:

Reviewer: ```suggestion
We'll also build our library so we can use its types from the dojo:
```

```
> =user-lib -build-file %/lib/json-test/hoon
```

Now we can ask Clay to build a mark conversion gate from a `%json` mark to our `%user` mark. We'll use a scry with a `%f` `care` which produces a static mark conversion gate:

```
> =json-to-user .^($-(json user:user-lib) %cf /===/json/user)
```

> Let's try convert our `$json` to a `$user` structure with our new mark conversion gate:

Reviewer: ```suggestion
Let's try converting our `$json` to a `$user` structure with our new mark conversion gate:
```

For the `name` field we've just formed a cell of `%a` and a list of `$json` strings, since a `$json` array is `[%a p=(list json)]`. Note we've separated this part into its own arm and wrapped the whole thing in a `|^` - a core with a `$` arm that's computed immediately. This is simply for readability - our structure here is quite simple but when dealing with deeply-nested `$json` structures or complex logic, having a single giant function can quickly become unwieldy.

#### Converting from `$json`

Here's our arm that converts `$json` to our `$user` structure:

```hoon
++  from-js
  =,  dejs:format
  |=  jon=json
  ^-  user

>   %.  jon

Reviewer 2: This is not strictly necessary, cause these runes are effectively a no-op. I'd scrap these three lines and replace it with  `^-  $-(json user)` cast, to make it clear that the `dejs` reparsers are higher order functions

For `joined`, we've used the [`+du`](/docs/hoon/reference/zuse/2d_6#dudejsformat) function, which converts a Unix seconds timestamp in a `$json` number to a `@da` (it's basically the inverse of the `+sect:enjs:format` we used earlier).

Notice how `+ot` takes in other `+dejs` functions in its argument. One of its arguments includes the `+at` function which itself takes in other `+dejs` functions. There are several `+dejs` functions like this that allow complex nested JSON structures to be decoded. For other examples of common `+dejs` functions like this, see the [More `+dejs`](#more-dejs) section below.

There are dozens of different functions in [`+dejs:format`](/docs/hoon/reference/zuse/2d_6) that will cover a great many use cases. If there isn't a `+dejs` function for a particular case, you can also just write a custom function - it just has to take `$json`. Note there's also the [`+dejs-soft:format`](/docs/hoon/reference/zuse/2d_7) functions - these are similar to `+dejs` functions except they produce `unit`s rather than simply crashing if decoding fails.

> ## More `+dejs`

Reviewer 2: Nothing on `+su`? Imo one of the most important decoders, because there's a lot of pressure on JSON objects to be somewhat stringly-typed (keys must be strings, strings don't break pointer equality unexpectedly).

```
> =usr (json-to-user jon)
> usr
[ username='john456'
  name=[first='John' mid='William' last='Smith']
  joined=~2021.9.12..09.47.58
  email='john.smith@example.com'
]
```

Now let's try the other direction. We'll again scry Clay to build a static mark conversion gate, this time _from_ `%user` _to_ `%json` rather than the reverse:

```
> =user-to-json .^($-(user:user-lib json) %cf /===/user/json)
```

Let's test it out by giving it our `$user` data:

```
> (user-to-json usr)
[ %o
    p
  { [p='email' q=[%s p='john.smith@example.com']]
    [p='name' q=[%a p=~[[%s p='John'] [%s p='William'] [%s p='Smith']]]]
    [p='username' q=[%s p='john456']]
    [p='joined' q=[%n p=~.1631440078]]
  }
]
```

Finally, let's see how that looks as JSON encoded in text:

```
> (crip (en-json:html (user-to-json usr)))
'{"joined":1631440078,"username":"john456","name":["John","William","Smith"],"email":"john.smith@example.com"}'
```

Usually (though not in all cases) these mark conversions will be performed implicitly by Gall or Eyre and you'd not deal with the mark conversion gates directly, but it's still informative to see them work explicitly.

> ## Further reading

Reviewer 2: Worth linking directly to `scot` imo, given its prevalence in JSON conversion.
